content/blog/angular-extend-class/index.md

```diff
@@ -0,0 +1,842 @@
+
+```typescript
+@Injectable()
+class BaseComponent implements OnInit {
+  ngOnInit() {
+    console.log('I AM BASE COMPONENT');
+  }
+}
+
+@Component({
+  selector: 'app-root',
+  template: `
+    <p>Test</p>
+  `,
+})
+class AppComponent extends BaseComponent {
+}
+```
+
+That said, [while using `@Injectable` is explicitly supported](https://github.com/angular/angular/issues/41229#issuecomment-800310757), it's very hacky to use this in place of `@Component`. This is because Angular's `@Injectable`s do not support lifecycle methods without a `Component` that extends it.
+
+As such, we'll be sticking to the `abstract` class solution.
+
+# Overwriting Lifecycle Methods {#lifecycle-methods}
+
+If you recall from our quick overview of what a base class does, you can replace the base class implementation of both methods and properties.
+
+The same is true for lifecycle methods, since they're just a type of method on the component class instance.
+
+```typescript
+@Component({
+  template: ''
+})
+abstract class BaseComponent implements OnInit {
+  ngOnInit() {
+    console.log('I AM BASE COMPONENT');
+  }
+}
+
+@Component({
+  selector: 'app-root',
+  template: `
+    <p>Test</p>
+  `,
+})
+class AppComponent extends BaseComponent implements OnInit {
+  override ngOnInit() {
+    console.log("And I am the AppComponent")
+  }
+}
+```
+
+The downside here, however, is that `ngOnInit` on `AppComponent` will no longer call the `BaseComponent`'s `ngOnInit` logic. After all, what if you wanted to simply _add_ behavior to `ngOnInit`, rather than replace it entirely?
+
+Luckily, we can use the `super` keyword to refer to the base class instance and call the original method inside of the overwritten method:
+
+```typescript
+@Component({
+  selector: 'app-root',
+  template: `
+    <p>Test</p>
+  `,
+})
+class AppComponent extends BaseComponent implements OnInit {
+  override ngOnInit() {
+    // This will log `I AM BASE COMPONENT`
+  	super.ngOnInit();
+    console.log("And I am the AppComponent")
+  }
+}
+```
+
+# Use Dependency Injection with your extended class {#dependency-injection}
+
+While you're able to use global `window` variable in a browser environment, if you're attempting to use `window` in a [server-side rendered](/posts/what-is-ssr-and-ssg) Angular application, it will throw an error.
+
+```
+window is not defined
+```
+
+To solve this problem you can use Angular's dependency injection to inject an instance of `document` to `BaseComponent`, and get access to the `window` through `defaultView` that way:
+
+```javascript
+@Component({
+  template: ''
+})
+abstract class BaseComponent implements OnInit, OnDestroy {
+  window!: Window;
+  constructor(@Inject(DOCUMENT) private document: Document) {
+    this.window = document.defaultView!;
+  }
+}
+```
+
+Because of this, this is the recommended way to get access to both the `document` and `window` instance inside of an Angular component, even if it's a non-SSR app.
```

Comment:
question: Where have you seen this as a recommended way of accessing the `window` object?

Reply:
issue: My recommendation is to never access the `window` object directly in any Angular-specific class. Instead, I would suggest adding an `InjectionToken` providing a data structure such as:

```typescript
interface ViewportDimensions {
  readonly innerHeight: number
  readonly innerWidth: number;
}
```

We would provide it something like this:
```typescript
const viewportDimensionsToken = new InjectionToken<ViewportDimensions>('viewportDimensionsToken', {
  providedIn: 'platform',
  factory: (): ViewportDimensions => ({
    get innerHeight() {
      return window.innerHeight;
    },
    get innerWidth() {
      return window.innerWidth;
    },
  }),
});
```